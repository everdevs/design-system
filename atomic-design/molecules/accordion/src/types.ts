import React from "react";
import { PropsWithTheme } from "@evernest/theme";
import { SpringConfig } from "react-spring";

export type AccordionElement = HTMLDivElement;

export interface AccordionProps extends PropsWithTheme {
	id?: string;
	title: string;
	dark?: boolean;
	springConfig?: SpringConfig;
	headerComponent?: keyof JSX.IntrinsicElements | React.ComponentType<any>;
}

export interface StyledPanelProps extends PropsWithTheme {
	expanded: boolean;
}

export interface StyledAccordionProps extends PropsWithTheme<AccordionElement> {}
