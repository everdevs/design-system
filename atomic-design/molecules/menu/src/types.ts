import { PropsWithTheme } from "@evernest/theme";

export type MenuElement = HTMLDivElement;

export enum MenuDirection {
	vertical = "vertical",
	horizontal = "horizontal",
}

export interface StyledMenuProps extends PropsWithTheme<MenuElement> {
	/** Add types */
	direction?: MenuDirection;
}
