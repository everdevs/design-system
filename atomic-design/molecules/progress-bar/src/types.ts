import { PropsWithTheme } from "@evernest/theme";

export type ProgressBarElement = HTMLDivElement;

export interface StyledProgressBarProps extends PropsWithTheme<ProgressBarElement> {}

export interface ProgressBarProps extends PropsWithTheme<ProgressBarElement> {
	ariaLabel: string;
	current: number;
	segments: number;
}
