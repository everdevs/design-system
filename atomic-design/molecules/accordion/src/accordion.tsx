import { css } from "@emotion/core";
import styled from "@emotion/styled";
import { Icon, Size } from "@evernest/icon";
import { PropsWithTheme } from "@evernest/theme";
import React from "react";
import { animated, useSpring } from "react-spring";
import { useMeasure } from "react-use";
import { AccordionElement, AccordionProps, StyledAccordionProps, StyledPanelProps } from "./types";

/* @todo: revisit border style with token refactor */
export const StyledAccordion = styled("div", {
	shouldForwardProp: (propName: string) => !["theme"].includes(propName),
})<StyledAccordionProps & { dark: boolean }>`
	${({ theme: { palette }, dark }) => css`
		border-bottom: 1px solid ${dark ? palette.white.css : palette.grey.css};
	`};
`;

export const StyledButton = styled.button`
	appearance: none;
	background: none;
	border-radius: 0;
	border: none;
	color: currentColor;
	font-family: inherit;
	font-size: inherit;
	padding: 0 0 var(--spacing-xs);
	width: 100%;
`;

export const StyledInnerButtonWrapper = styled.div`
	display: flex;
	height: 100%;
	overflow: hidden;
	text-align: left;
`;

export const StyledPanel = styled.div<StyledPanelProps>`
	padding-bottom: var(--spacing-xs);
	padding-left: var(--spacing-s);
	> :first-child {
		margin-top: 0;
	}

	> :last-child {
		margin-bottom: 0;
	}

	${({ theme: { mq }, expanded }) => css`
		visibility: ${expanded ? "visible" : "hidden"};
		@media ${mq.l} {
			padding-left: var(--spacing-m);
		}
	`};
`;

export const StyledAnimatedPanelWrapper = styled(animated.div)`
	overflow: hidden;
`;

export const StyledIconWrapper = styled(animated.div)<PropsWithTheme & { dark: boolean }>`
	display: inline-flex;
	min-width: var(--spacing-s);
	${({ theme: { mq, palette }, dark }) => css`
		color: ${dark ? palette.white.css : palette.grey.css};
		@media ${mq.l} {
			min-width: var(--spacing-m);
		}
	`};
`;

export const StyledAnimatedIconWrapper = styled(animated.span)`
	display: inline-flex;
`;

export const Accordion = React.forwardRef<AccordionElement, AccordionProps>(
	({ id, children, dark, title, headerComponent, springConfig, ...props }, ref) => {
		const [expanded, setExpanded] = React.useState(false);
		const [panelBottomPadding, setPanelBottomPadding] = React.useState(0);

		const handleClick = () => setExpanded(!expanded);

		const panelId = `${id}-panel`;

		const [useMeasureRef, { height }] = useMeasure<HTMLDivElement>();

		const springPanelProps = useSpring({
			config: springConfig,
			height: expanded ? height + panelBottomPadding : 0,
		});

		const { z } = useSpring({
			config: springConfig,
			z: expanded ? 0.125 : 0,
		});

		React.useEffect(() => {
			const panelEl = document.getElementById(panelId);
			/* istanbul ignore next */
			if (panelEl) {
				const paddingBottom = getComputedStyle(panelEl).paddingBottom;
				setPanelBottomPadding(parseInt(paddingBottom));
			}
		}, [panelId, setPanelBottomPadding]);

		const HeaderComponent = React.useMemo(() => headerComponent, [headerComponent]);

		return (
			<StyledAccordion {...props} dark={dark} ref={ref}>
				<HeaderComponent>
					<StyledButton
						aria-controls={panelId}
						aria-expanded={expanded}
						id={id}
						onClick={handleClick}
					>
						<StyledInnerButtonWrapper>
							<StyledIconWrapper dark={dark}>
								<StyledAnimatedIconWrapper
									style={{
										transform: z.interpolate(
											value => `rotate3d(0,0,1,${value}turn)`
										),
									}}
								>
									<Icon aria-hidden="true" icon="close" size={Size.medium} />
								</StyledAnimatedIconWrapper>
							</StyledIconWrapper>
							<span data-test-id="styled-inner-button-wrapper-label">{title}</span>
						</StyledInnerButtonWrapper>
					</StyledButton>
				</HeaderComponent>
				<StyledAnimatedPanelWrapper style={springPanelProps}>
					<StyledPanel
						expanded={expanded}
						ref={useMeasureRef}
						aria-labelledby={id}
						id={panelId}
						role="region"
					>
						{children}
					</StyledPanel>
				</StyledAnimatedPanelWrapper>
			</StyledAccordion>
		);
	}
);

Accordion.defaultProps = {
	dark: false,
	headerComponent: "div",
	springConfig: undefined,
};
