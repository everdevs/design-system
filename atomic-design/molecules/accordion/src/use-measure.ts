import React from "react";

export interface Measurements {
	contentRect: Pick<
		DOMRectReadOnly,
		"x" | "y" | "top" | "left" | "right" | "bottom" | "height" | "width"
	>;
	borderBoxSize: ResizeObserverSize;
	contentBoxSize: ResizeObserverSize;
}

const intialState = {
	contentRect: {
		x: 0,
		y: 0,
		width: 0,
		height: 0,
		top: 0,
		right: 0,
		bottom: 0,
		left: 0,
	},
	borderBoxSize: {
		inlineSize: 0,
		blockSize: 0,
	},
	contentBoxSize: {
		inlineSize: 0,
		blockSize: 0,
	},
};

/* @todo: revisit testing when ResizeObserver is better implemented */
/* istanbul ignore next */
const useMeasure = <T extends Element = HTMLDivElement>(): [
	React.MutableRefObject<T>,
	Measurements
] => {
	const [bounds, setBounds] = React.useState<Measurements>(intialState);
	const ref = React.useRef<T>();

	React.useLayoutEffect(() => {
		let animationFrameId: number;

		const measure: ResizeObserverCallback = ([entry]) => {
			animationFrameId = window.requestAnimationFrame(() => {
				setBounds({
					contentRect: entry.contentRect,
					borderBoxSize: entry.borderBoxSize[0],
					contentBoxSize: entry.contentBoxSize[0],
				});
			});
		};

		const resizeObserver = new ResizeObserver(measure);
		resizeObserver.observe(ref.current);

		return () => {
			window.cancelAnimationFrame(animationFrameId);
			resizeObserver.disconnect();
		};
	}, []);

	return [ref, bounds];
};

export default useMeasure;
