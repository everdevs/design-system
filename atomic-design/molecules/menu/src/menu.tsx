import { css } from "@emotion/react";
import styled from "@emotion/styled";
import React from "react";
import { MenuDirection, MenuElement, StyledMenuProps } from "./types";

export const StyledMenuTpl = styled.nav<StyledMenuProps>`
	display: flex;
	${({ direction }) => css`
		flex-direction: ${direction === MenuDirection.vertical ? "column" : "row"};
	`};
`;

export const Menu = React.forwardRef<MenuElement, StyledMenuProps>((props, ref) => (
	<StyledMenuTpl {...props} ref={ref} />
));
