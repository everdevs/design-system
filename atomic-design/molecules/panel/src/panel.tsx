import styled from "@emotion/styled";
import { css, Global } from "@emotion/core";
import React from "react";
import { Anchor, PanelElement, StyledBackdropProps, StyledPanelProps } from "./types";

export const StyledBackdrop = styled.div<StyledBackdropProps>`
	position: fixed;
	z-index: 100;
	top: 0;
	right: 0;
	bottom: 0;
	left: 0;
	opacity: 0;
	visibility: hidden;
	pointer-events: none;
	transition-duration: 0.3s;
	transition-property: opacity, visibility;
	${({ theme: { palette } }) =>
		css`
			background: ${palette.black.css};
		`};
	${({ open }) =>
		open &&
		css`
			visibility: visible;
			pointer-events: all;
			opacity: 0.7;
		`};
`;

export const StyledPanel = styled.aside<StyledPanelProps>`
	position: fixed;
	z-index: 100;
	transition-duration: 0.3s;
	transition-property: transform;
	overflow: auto;
	${({ theme: { palette } }) => css`
		background: ${palette.white.css};
	`};
	${({ anchor, width, open }) => {
		switch (anchor) {
			case Anchor.bottom: {
				return css`
					bottom: 0;
					left: 0;
					right: 0;
					transform: translate3d(0, ${open ? 0 : "100%"}, 0);
				`;
			}
			case Anchor.left: {
				return css`
					bottom: 0;
					left: 0;
					top: 0;
					width: ${width};
					max-width: calc(100vw - 80px);
					transform: translate3d(${open ? 0 : "-100%"}, 0, 0);
				`;
			}
			case Anchor.right: {
				return css`
					bottom: 0;
					top: 0;
					right: 0;
					width: ${width};
					max-width: calc(100vw - 80px);
					transform: translate3d(${open ? 0 : "100%"}, 0, 0);
				`;
			}
			case Anchor.top: {
				return css`
					top: 0;
					left: 0;
					right: 0;
					transform: translate3d(0, ${open ? 0 : "-100%"}, 0);
				`;
			}
			default: {
				break;
			}
		}
	}};
`;

const blockScroll = css`
	body {
		overflow: hidden;
		height: 100vh;
	}
`;

export const Panel = React.forwardRef<PanelElement, StyledPanelProps>(
	({ open, onClose, ...props }, ref) => {
		const handleClick = React.useCallback(() => {
			open && onClose();
		}, [open, onClose]);
		return (
			<>
				{open && <Global styles={blockScroll} />}
				<StyledBackdrop open={open} onClick={handleClick} />
				<StyledPanel {...props} open={open} ref={ref} />
			</>
		);
	}
);

Panel.defaultProps = {
	width: "295px",
};
