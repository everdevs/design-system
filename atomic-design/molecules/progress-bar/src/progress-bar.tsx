import styled from "@emotion/styled";
import React from "react";
import { ProgressBarElement, ProgressBarProps, StyledProgressBarProps } from "./types";
import { animated, useSpring } from "react-spring";
import { css } from "@emotion/react";

const StyledProgressBar = styled(animated.div)<StyledProgressBarProps>`
	height: 8px;
	${({ theme: { palette } }) => css`
		background: ${palette.green.css};
	`};
`;

export const ProgressBar = React.forwardRef<ProgressBarElement, ProgressBarProps>((props, ref) => {
	const { label, current, segments, ...rest } = props;
	const percentage = Math.ceil((current / segments) * 100);
	const animatedStyles = useSpring({ width: `${percentage > 100 ? 100 : percentage}%` });
	return (
		<StyledProgressBar
			aria-valuemin={0}
			aria-valuemax={segments}
			aria-valuenow={current}
			aria-valuetext={label}
			ref={ref}
			role="progressbar"
			style={(animatedStyles as unknown) as React.CSSProperties}
			{...rest}
		/>
	);
});
