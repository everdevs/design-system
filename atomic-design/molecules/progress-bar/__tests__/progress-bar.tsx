import { createTree, withTheme } from "@evernest/dev-helpers";
import { theme } from "@evernest/theme";
import React from "react";
import { ProgressBar, ProgressBarProps } from "../src";

const ThemedProgressBar = withTheme<ProgressBarProps>(ProgressBar, theme);

test("ProgressBar matches the snapshot", () => {
	const tree = createTree(<ThemedProgressBar label="Step 1 of 3" current={2} segments={3} />);
	expect(tree).toMatchSnapshot();
});

test("ProgressBar should clamp to 100", () => {
	const tree = createTree(<ThemedProgressBar label="Step 4 of 3" current={4} segments={3} />);
	expect(tree).toMatchSnapshot();
});
