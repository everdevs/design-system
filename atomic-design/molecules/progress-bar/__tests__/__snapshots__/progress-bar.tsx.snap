// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar matches the snapshot 1`] = `
<div
  aria-valuemax={3}
  aria-valuemin={0}
  aria-valuenow={2}
  aria-valuetext="Step 1 of 3"
  className="css-1l8ew5k"
  role="progressbar"
  style={
    Object {
      "width": "67%",
    }
  }
/>
`;

exports[`ProgressBar should clamp to 100 1`] = `
<div
  aria-valuemax={3}
  aria-valuemin={0}
  aria-valuenow={4}
  aria-valuetext="Step 4 of 3"
  className="css-1l8ew5k"
  role="progressbar"
  style={
    Object {
      "width": "100%",
    }
  }
/>
`;
